src/main/scala/mesosphere/marathon/core/storage/migration/MigrationTo1_3_PersistentStore.scala

```diff
@@ -0,0 +1,157 @@
+      oldRepo.all().mapAsync(Int.MaxValue) { value =>
+        newRepo.store(value)
+      }.runFold(0) { case (acc, _) => acc + 1 }
+    }
+    await {
+      oldRepo.ids().mapAsync(Int.MaxValue) { id =>
+        oldRepo.delete(id)
+      }.runWith(Sink.ignore).asTry
+    }
+    migrated
+  }
+
+  def migrateVersionedRepo[Id, T](
+    oldRepo: VersionedRepository[Id, T],
+    newRepo: VersionedRepository[Id, T]): Future[Int] = async {
+    val oldVersions = oldRepo.ids().flatMapConcat { id =>
+      oldRepo.versions(id).mapAsync(Int.MaxValue) { version =>
+        oldRepo.getVersion(id, version)
+      }.collect { case Some(value) => value }
+    }.mapAsync(1) { value =>
+      oldRepo.storeVersion(value)
+    }.runFold(0) { case (acc, _) => acc + 1 }
+
+    val currentVersions = oldRepo.all().mapAsync(1) { value =>
+      newRepo.store(value)
+    }.runFold(0) { case (acc, _) => acc + 1 }
+
+    val result = await(oldVersions) + await(currentVersions)
+    await(oldRepo.ids().mapAsync(Int.MaxValue)(oldRepo.delete).runWith(Sink.ignore).astTry)
+    result
+  }
+
+  def migrateTasks(legacyStore: LegacyStorageConfig, taskRepository: TaskRepository): Future[(String, Int)] = {
+    val oldRepo = TaskRepository.legacyRepository(legacyStore.entityStore[MarathonTaskState])
+    migrateRepo(oldRepo, taskRepository).map("tasks" -> _)
+  }
+
+  def migrateDeployments(
+    legacyStore: LegacyStorageConfig,
+    deploymentRepository: DeploymentRepository): Future[(String, Int)] = {
+    val oldRepo = DeploymentRepository.legacyRepository(legacyStore.entityStore[DeploymentPlan])
+    migrateRepo(oldRepo, deploymentRepository).map("deployment plans" -> _)
+  }
+
+  def migrateTaskFailures(
+    legacyStore: LegacyStorageConfig,
+    taskFailureRepository: TaskFailureRepository): Future[(String, Int)] = {
+    val oldRepo = TaskFailureRepository.legacyRepository(legacyStore.entityStore[TaskFailure])
+    migrateVersionedRepo(oldRepo, taskFailureRepository).map("task failures" -> _)
+  }
+
+  def migrateGroups(
+    legacyStore: LegacyStorageConfig,
+    groupRepository: GroupRepository): Future[(String, Int)] = {
+    val oldAppRepo = AppRepository.legacyRepository(legacyStore.entityStore[AppDefinition], legacyStore.maxVersions)
+    val oldRepo = GroupRepository.legacyRepository(legacyStore.entityStore[Group], legacyStore.maxVersions, oldAppRepo)
+
+    val result = oldRepo.rootVersions().mapAsync(Int.MaxValue) { version =>
+      oldRepo.rootVersion(version)
+    }.collect {
+      case Some(root) => root
+    }.concat { Source.fromFuture(oldRepo.root()) }.mapAsync(1) { root =>
+      // we store the roots one at a time with the current root last,
+      // adding a new app version for every root (for simplicity)
+      groupRepository.storeRoot(root, root.apps.values.toVector, Nil)
+    }.runFold(0) { case (acc, _) => acc + 1 }.map("root versions" -> _)
+    await(oldRepo.ids().mapAsync(Int.MaxValue)(oldRepo.delete).runWith(Sink.ignore).asTry)
```

Comment:
You do not wait for the delete operation in the resulting future. Surround with async and await `result` as well.